ZukoAzula2016/src/org/team1540/zukoazula/AutonomousBase.java

```diff
@@ -0,0 +1,159 @@
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+
+import ccre.channel.BooleanInput;
+import ccre.channel.FloatInput;
+import ccre.cluck.Cluck;
+import ccre.instinct.AutonomousModeOverException;
+import ccre.instinct.InstinctModeModule;
+import ccre.log.Logger;
+import ccre.rconf.RConf;
+import ccre.rconf.RConf.Entry;
+import ccre.rconf.RConfable;
+import ccre.tuning.TuningContext;
+
+public abstract class AutonomousBase extends InstinctModeModule {
+
+    public AutonomousBase(String modeName) {
+        super(modeName);
+    }
+
+    @Override
+    protected void autonomousMain() throws InterruptedException, AutonomousModeOverException {
+        try {
+            runAutonomous();
+        } finally {
+            Autonomous.allMotors.set(0);
+            Autonomous.fire.set(false);
+            Autonomous.warmup.set(false);
+            Autonomous.intakeArm.set(0);
+        }
+    }
+
+    protected abstract void runAutonomous() throws InterruptedException, AutonomousModeOverException;
+
+    protected void driveForTime(long time, float speed) throws AutonomousModeOverException, InterruptedException {
+        try {
+            Autonomous.allMotors.set(speed);
+            waitForTime(Math.abs(time));
+        } finally {
+            Autonomous.allMotors.set(0);
+        }
+    }
+
+    protected void driveDistance(float dist, float speed) throws AutonomousModeOverException, InterruptedException {
+        try {
+            float start = Autonomous.driveEncoder.get();
+            Autonomous.allMotors.set(dist > 0 ? speed : -speed);
+            if (dist > 0) {
+                waitUntilAtLeast(Autonomous.driveEncoder, start + dist);
+            } else {
+                waitUntilAtMost(Autonomous.driveEncoder, start + dist);
+            }
+        } finally {
+            Autonomous.allMotors.set(0);
+        }
+    }
+
+    protected void turnAngle(int angle, int speed) throws AutonomousModeOverException, InterruptedException {
+        try {
+            float start = Autonomous.absoluteYaw.get();
+            Autonomous.turnMotors.set(angle > 0 ? speed : -speed);
+            if (angle > 0) {
+                waitUntilAtLeast(Autonomous.absoluteYaw, start + angle);
+            } else {
+                waitUntilAtMost(Autonomous.absoluteYaw, start + angle);
+            }
+        } finally {
+            Autonomous.allMotors.set(0);
+        }
+    }
+
+    protected void startWarmup() throws AutonomousModeOverException, InterruptedException {
+        Autonomous.warmup.set(true);
+    }
+
+    protected void fire(long time) throws AutonomousModeOverException, InterruptedException {
+        Autonomous.fire.set(true);
+        waitForTime(time);
+        Autonomous.fire.set(false);
+        Autonomous.warmup.set(false);
+    }
+
+    protected void intakeArmLower(float speed) throws AutonomousModeOverException, InterruptedException {
+        Autonomous.intakeArm.set(-Math.abs(speed));
```

Comment:
Perhaps it would be better to have one `moveIntakeArm(float speed)` and document the polarity...